src/mgmt/srv_pool.c

```diff
@@ -162,6 +162,8 @@ ds_mgmt_tgt_pool_create_ranks(uuid_t pool_uuid, char *tgt_dev,
 		D_DEBUG(DB_TRACE, "fill ranks %d idx %d "DF_UUID"\n",
 			tc_out_ranks[i], idx, DP_UUID(tc_out_uuids[i]));
 	}
+	D_FREE(tc_out->tc_tgt_uuids.ca_arrays);
+	D_FREE(tc_out->tc_ranks.ca_arrays);
```

Comment:
these two will be freed internally by cart when crt_hg_req_destroy -> HG_Free_output() -> proc func(). no?

Reply:
No! This was my assumption also. But unfortunately Mercurry free only structs which are decoded previously. Those arrays are for response and user should free them manually.

Reply:
I agree this was a bit surprise for me also. But we have completion callback which is responsible to deallocated all resources from reply. We should just care about this. In case of dynamically generated serialization functions they don't allocate anything from encoding path. So, they need not release anything. From this point of view pipeline is correct.

Maybe we should think about how to care about those resources in our DAOS code more transparent but I don't have a good solution for this yet.

Reply:
Just to clarify.
For receiving RPC the pipeline is following:
- CRT_PROC_DECODE (in structures) -> CRT_PROC_ENCODE (out structures) and send reply -> CRT_PROC_FREE (in structures)
For sending RPC the pipeline is following:
- CRT_PROC_ENCODE (in structures) -> send RPC and wait for reply, CRT_PROC_DECODE (out structures) -> CRT_PROC_FREE (out structures)

So, Mercurry free only structures which where previously decoded. They can be as in as out in case of rpc direction.

Reply:
As I mentioned before this was surprise for me also. I expected the flow you described above. But something changed in Mercury or it was misunderstanding from begin the work flow now is different. I'm not ready to change this flow in Mercury now in this patch. Probably later when memory leak blocking testing and customer will be resolved we can discover this and change work flow according our understanding.